tools/unicode_properties_parse/format_width_estimate_intervals.py

```diff
@@ -0,0 +1,184 @@
+        range_u(0x2E80, 0x303E),
+        range_u(0x3040, 0xA4CF),
+        range_u(0xAC00, 0xD7A3),
+        range_u(0xF900, 0xFAFF),
+        range_u(0xFE10, 0xFE19),
+        range_u(0xFE30, 0xFE6F),
+        range_u(0xFF00, 0xFF60),
+        range_u(0xFFE0, 0xFFE6),
+        range_u(0x1F300, 0x1F64F),
+        range_u(0x1F900, 0x1F9FF),
+        range_u(0x20000, 0x2FFFD),
+        range_u(0x30000, 0x3FFFD),
+    ]
+
+    table = table_u()
+    for rng in std_wide_ranges_cpp20:
+        table.fill_range(rng, width_u.is_2)
+
+    return table
+
+
+# Read data from "EastAsianWidth.txt".
+# The latest version can be found at:
+# https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt
+# The current implementation works for:
+# https://www.unicode.org/Public/15.0.0/ucd/EastAsianWidth.txt
+# To make this function work, the file should not contain a BOM.
+def read_from(source: TextIO) -> table_u:
+    table = table_u()
+
+    # "The unassigned code points in the following blocks default to "W":"
+    default_wide_ranges = [
+        range_u(0x4E00, 0x9FFF),
+        range_u(0x3400, 0x4DBF),
+        range_u(0xF900, 0xFAFF),
+        range_u(0x20000, 0x2FFFD),
+        range_u(0x30000, 0x3FFFD),
+    ]
+    for rng in default_wide_ranges:
+        table.fill_range(rng, width_u.is_2)
+
+    # Read explicitly assigned ranges.
+    # The lines that are not empty or pure comment are uniformly of the format "HEX(..HEX)?;(A|F|H|N|Na|W) #comment".
+    def get_width(str: str):
+        if str == "F" or str == "W":
+            return width_u.is_2
+        else:
+            assert str == "A" or str == "H" or str == "N" or str == "Na"
+            return width_u.is_1
+
+    reg = re.compile(r"([0-9A-Z]+)(\.\.[0-9A-Z]+)?;(A|F|H|N|Na|W) *#.*")
+    for line in source:
+        line = line.strip()
+        if line and not line.startswith("#"):
+            match = reg.fullmatch(line)
+            assert match, "invalid line"
+            from_val = int(match.group(1), base=16)
+            width = get_width(match.group(3))
+            if match.group(2):
+                # range (HEX..HEX)
+                to_val = int(match.group(2)[2:], base=16)
+                table.fill_range(range_u(from_val, to_val), width)
+            else:
+                # single character (HEX)
+                table.fill_range(range_u(from_val), width)
+
+    return table
+
+
+def get_table_cpp23(source: TextIO) -> table_u:
+    table = read_from(source)
+
+    # Override with ranges specified by the C++ standard.
+    std_wide_ranges_cpp23 = [
+        range_u(0x4DC0, 0x4DFF),
+        range_u(0x1F300, 0x1F5FF),
+        range_u(0x1F900, 0x1F9FF),
+    ]
+
+    for rng in std_wide_ranges_cpp23:
+        table.fill_range(rng, width_u.is_2)
+
+    return table
+
+
+def main():
+    print("Old table:")
+    old_table = get_table_cpp20()
+    old_table.print_intervals()
+
+    print("\nNew table:\nInput path for EastAsianWidth.txt: ", end="")
+    path = input()

```

Comment:
we can just load this like the other generators (I forget if they use parameters or just load from next to the script)